cpu/native/include/periph_cpu.h

```diff
@@ -63,6 +59,10 @@ extern "C" {
  */
 #define GPIO_PIN(port, pin) (gpio_t)((port << GPIO_PORT_SHIFT) | pin)
 
+/* GPIO configuration only if the module is available (=Linux) */
```

Comment:
I'm unsure this is the right approach here:

The mock implementation is distinct from the native implementation, so reusing the same settings for both is confusing.

What is wrong with GPIO_PIN(1,3) == GPIO_PIN(0, 3)? The hardware that is being mocked could just only have one port. Even with the current defines, GPIO_PIN(0, 1<<24) would be identical to GPIO_PIN(1, 0); aliasing is only avoided if the mock module either established some boundaries or chose to use a two-int sized (eg. struct) type as its gpio_t (which I'd welcome as a test to see whether any piece of the code tacitly assumes that gpio_t can be cast around).

Reply:
Each GPIO "identifier" should be unique.
If I want to mock STM32 GPIOs for instance, GPIO_PIN(PORT_A, 1) is not the same than GPIO_PIN(PORT_B, 1). If we want to mock GPIOs, it should be close to real world.
But it is hard to define some boundaries as it is hard to predict what will be mocked.

Moreover, if you want to perform a switch case, if the macro returns twice same value it leads to such kind of error:
```shell
pf_positional_actuators.cpp: In function ‘void* cogip::pf::actuators::positional_actuators::_gpio_handling_thread(void*)’:
pf_positional_actuators.cpp:218:9: error: duplicate case value
  218 |         case pin_limit_switch_left_arm_lift_bottom:
      |         ^~~~
pf_positional_actuators.cpp:202:9: note: previously used here
  202 |         case pin_limit_switch_central_lift_bottom:
      |         ^~~~
pf_positional_actuators.cpp:233:9: error: duplicate case value
  233 |         case pin_sensor_pump_left:
      |         ^~~~
pf_positional_actuators.cpp:214:9: note: previously used here
  214 |         case pin_limit_switch_right_arm_lift_top:
      |         ^~~~
```

Reply:
The GPIO pin and port arguments, though theoretically generic, are both integer, and thus the only way to avoid equality of valid GPIOs are to either decide to subset them at check time (say, use 16 bits for pin and 16 bit for port number, and enforce that no other pins are valid), or to use a larger than default gpio_t.

To some extent, the bug here is that arbitrary pins are created in the first place -- any time input to GPIO_PIN is provided, the application needs to know that that pin exists (or rely on the verify function, which in the mock case is not accurately representing that there is really just one port). Usually the pins are configurable per board, and for native they would just all be mapped to the 0 port.

If the gpio_mock module is supposed to be used for arbitrary pin setup emulation (as is suggested by the presence of the weak attributes; a comment similar to the one on the i2c side could clarify that), I think that the best route would be to `typedef struct { unsigned int port; unsigned in pin; bool is_undef; } gpio_t;` and `#define GPIO_PIN` accordingly (possibly through a static inline function because C is odd about where struct literals can be used).
Alternatively, copy what is done in the gpio_linux case and pick some grouping of the bits what is for registers and what is for pins. But in that case, that choice (including the concrete shift) becomes part of the public API for creating the per-application gpio_… mock functions, and they need to pick their arguments apart again using the same shift.

Semi-relatedly, I'm not sure doing a switch-case on pins is futureproof -- there exists an `gpio_is_equal` function that may change its implementation in the future.

Reply:
@chrysn I now understand what you mean and I agree. 👍
However, it's not entirely straightforward, because if I change `gpio_t` to something other (struct, pointer) than the default uint32_t type, it results in a build error if I mix with a pcf875x gpio expander for example.

I think it's not as simple as I thought at first and it would take more time to get it right.
And it is always possible to define my own GPIO_PIN macro in the context of the application or my own board, in `periph_conf.h` for instance.

I therefore propose to abandon this commit. Are you OK with that ?

Reply:
uh `pcf857x` should not (ab)use `gpio_t` like that. It expects a plain integer (index of the GPIO), using `gpio_t` here will just lead to chaos.

Reply:
```C
/** conversion of (port x : pin y) to a pin number */
#define PCF857X_GPIO_PIN(x,y)  ((gpio_t)((x << 3) | y))
```

All the `pcf857x` API is modeled on `periph_gpio`. As mainly using stm32 and native, planets were aligned but following @chrysn implementation, I confirm, it leads to chaos in my build due to pcf857x. :laughing: 

https://github.com/RIOT-OS/RIOT/blob/f048714418ec8097f80290423b1104380286cb38/drivers/include/pcf857x.h#L487

It seems more safe to drop this commit for now, just let me know.

Reply:
Ben fixed this already in https://github.com/RIOT-OS/RIOT/pull/20435, so the option of doing the struct would be on the table again. Abandoning this is fine, but if you happen to have already some code around where you tried the struct approach, I'd love to see it here.

Reply:
@chrysn what do you think about that proposal for gpio mock in last commit ?